File: docs/chapter1/section3/src/2-1_json-server.go
package main

import (
	"net/http"

	"github.com/labstack/echo/v4"
)

// JSONで返すための構造体を定義
type jsonData struct {
	Number int
	String string
	Bool   bool
}

func main() {
	// Echoの新しいインスタンスを作成
	e := echo.New()

	// 「/json」というパスのエンドポイントを定義
	e.GET("/json", jsonHandler)

	// Webサーバーをポート番号8080で起動し、エラーが発生した場合はログにエラーメッセージを出力する
	e.Logger.Fatal(e.Start(":8080"))
}

func jsonHandler(c echo.Context) error {
	// レスポンスとして返す値を構造体として定義
	response := jsonData{
		Number: 10,
		String: "hoge",
		Bool:   false,
	}

	// HTTPステータスコードは200番で、構造体をJSONに変換してクライアントに返す
	return c.JSON(http.StatusOK, &response)
}

File: docs/chapter1/section3/src/2-1_json-server.rs
use axum::{routing::get, Json, Router};

#[tokio::main]
async fn main() {
	// 「/json」というパスのエンドポイントを定義
    let app = Router::new()
        .route("/json", get(json_handler));

	// Webサーバーをポート番号8080にバインドする
    let listener = tokio::net::TcpListener::bind("127.0.0.1:8080")
        .await
        .unwrap();

    println!("listening on {}", listener.local_addr().unwrap());

	// サーバーを起動する
    axum::serve(listener, app).await.unwrap();
}

// JSONで返すための構造体を定義
// 構造体を JSON に変換するためにserde::Serializeを導出する
#[derive(serde::Serialize)]
struct JsonData {
    number: i32,
    string: String,
    bool: bool,
}

async fn json_handler() -> Json<JsonData> {
	// レスポンスとして返す値を構造体として定義
    let res = JsonData {
        number: 10,
        string: String::from("hoge"),
        bool: false,
    };

	// 構造体をJSONに変換してクライアントに返す
    Json(res)
}